src/sonic-yang-models/yang-models/sonic-extension.yang

```diff
@@ -7,7 +7,54 @@ module sonic-extension {
 
     description "Extension yang Module for SONiC OS";
 
+    revision 2021-02-23 {
+        description "Adding extensions as per Sonic Yang Guidelines";
+    }
+
     revision 2019-07-01 {
         description "First Revision";
     }
+
+    /* For complete guide of using these extensions in SONiC yangs, refer
+    SONiC yang guidelines at
+    https://github.com/Azure/SONiC/blob/master/doc/mgmt/SONiC_YANG_Model_Guidelines.md */
+
+    extension db-name {
+        description "DB name, e.g. APPL_DB, CONFIG_DB";
+        argument "value";
+    }
+
+    extension key-delim {
+        description "Key delimeter, e.g. - |, :";
+        argument "value";
+    }
+
+    extension key-pattern {
+        description "Key pattern, e.g. - ACL_RULE|{aclname}|{rulename}";
+        argument "value";
+    }
+
+    extension map-leaf {
+        description
+           "Extension to indicate nested list and specify fields
+           mappings which are keys of inner list";
+        argument "value";
+    }
+
+    extension custom-validation {
+        description "Extension for registering custom validation handler.";
+        argument "handler";
+    }
+
+    extension dependent-on {
```

Comment:
Sometimes order of table creation matters from CVL perspective but may not be from sonic-backend. This extension is explained with example in Guidelines update PR (https://github.com/Azure/SONiC/pull/754).

Reply:
So sonic back-end does not need this. But why CVL needs it, if the dependency is expressed in the yang model using leaf-refs already.

Reply:
There can be cases where dependency can't be expressed with leaf-refs.
For ex. When Spanning tree mode is enabled, if any port has switchport configured, STP will be enabled on that port also. For this entry has to be added in STP table and STP_PORT table. Mgmt_framework performs this DB update in single transaction. In STP_PORT, there is a field "link_type" which can be configured only if STP mode is "rpvst". For CVL to validate this field has to know STP mode which is not yet committed in DB as create request for both STP table and STP_PORT is coming in same transaction. So CVL needs validation request in order where STP entry creation should be followed by STP_PORT entry creation. There is no leaf-ref relationship between STP table and STP_PORT table.

            list STP_PORT_LIST {
                key "ifname";
                leaf ifname {
                    type union {
                        type leafref {
                            path "/prt:sonic-port/prt:PORT/prt:PORT_LIST/prt:ifname";
                        }
                        type leafref {
                            path "/spc:sonic-portchannel/spc:PORTCHANNEL/spc:PORTCHANNEL_LIST/spc:name";
                        }
                    }
                }
                ...
                leaf link_type {
                    must "../../../STP/STP_LIST[keyleaf='GLOBAL']/mode='rpvst'" {
                        error-message "Configuration allowed in Rapid PVST mode";
                        error-app-tag stp-invalid;
                    }
                    type enumeration {
                        enum auto;
                        enum shared;
                        enum point-to-point;
                    }
                }
             }

Reply:
How sonic backend handles it today. BTW, how a developer of yang model will know, when to define dependent-on.

Reply:
I am not sure how backend handles this. STP feature developer can provide more details. I explained the issue with respect to mgmt-framework and CVL and this extension helps to resolve order of table updates. Developer who is specifying the MUST condition should be aware of dependency between the tables and accordingly define dependent-on.

Reply:
@praveen-li We had hours of discussion explaining why the "dependent-on" extension is needed with the very clear example of STP_PORT and STP_LIST. We clearly explained when doing bulk configurations when these two configuration come together STP_LIST has to be created first before STP_PORT else the entire bulk operation will fail. Also we explained the design of CVL where we do not have everything in memory and need to be in the right order for the configurations to be validated using CVL. No objections were raised during the meeting and we thought we clearly explained why this is needed until we see the above comment from you. Can you please let us know what more we can do to help here?

Reply:
In meeting, We decided that if order is incorrect config validation should fail.

Reply:
Since we support bulking from REST and gNMI, we can't enforce user to always provide in correct order. So to take care of ordering we need this extension. As far as I remember, we didn't agreed to fail validation in case of incorrect order.

Reply:
> Since we support bulking from REST and gNMI, we can't enforce user to always provide in correct order. So to take care of ordering we need this extension. As far as I remember, we didn't agreed to fail validation in case of incorrect order.

IMO, Supporting Bulking should be till REST and the in the Library. YANG Model constraints must be a representation of Sonic Back-End, where we do not have any constraint on order of tables. Because other way, depend-on is needed for many cases.